### PR TITLE
OME-TIFF: improve multi-file discussion

### PR DESCRIPTION
**Pull request by sbesson**
See https://trello.com/c/AvPFhuq9/119-improve-documentation-on-multi-file-ome-tiff-filesets

This PR builds on top of the discussion initiated on the Micro-Manager and ome-devel mainling list and reviews our OME-TIFF format specification document especially in the context of multiple files.

Main changes in this PR are the following:
- separate the OME-TIFF file structure discussion from the OME-TIFF specification page
- include discussion on multiple vs single file, BigTiff, partial metadata in the file structure page
- review the multi-file OME-TIFF specification to 1- use exclusively snippets of our tested public samples, 2- include more links to our schema documentation

As this question has been raised in multiple occasions, getting it right is crtiical

Open questions:
- `Embedded OME-XML metadata` is not optimal, we don't have any word to describe `Non-partial OME-XML metadata`
- `.ome.xml` snippets - I included the full `tiffcomment | xmlvalid` dump for documentation purpose but only use the first 50 lines. Worth coming back to copying the blocks inline ?
- other layout proposals are welcome

### CODE REVIEW

File: formats/ome-tiff/file-structure.txt [hunk 1]
```diff
@@ -61,16 +60,17 @@ number of planes are stored in each TIFF.
...
-manipulate existing TIFF IFDs. It has the same drawbacks as the ones
-highlighted above in the sense that this companion file needs to be
-preserved as part of the fileset to prevent metadata loss.
+Since the :doc:`2011-06 version</schemas/june-2011>` of the OME-XML schema, it
+is possible to store partial OME-XML metadata blocks in some or all of the
+TIFF files pointing to a master file containing the full OME-XML metadata.
+The master file can be either another OME-TIFF file or a companion OME-XML
+file.
+
+Using a companion OME-XML file allows to easily append information that caz
```
Comment (mtbc) [inline formats/ome-tiff/file-structure.txt]: can?

**Comment by hflynn** [comment]
Staged at https://www.openmicroscopy.org/site/support/ome-model-staging/ome-tiff/file-structure.html and other pages at https://www.openmicroscopy.org/site/support/ome-model-staging/index.html#ome-tiff

File: formats/ome-tiff/file-structure.txt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+The following section discusses various considerations when designing
+OME-TIFF filesets distributed over multiple files.
+
+Single versus multiple files
+----------------------------
+
+Splitting a fileset across multiple files can have advantages in terms of
+acquisition but also processing purposes. The OME-TIFF file format can support
+any image organization. However, using one TIFF file per timepoint per channel with the focal planes for that timepoint and channel stored sequentially within the TIFF makes for very easy :ref:`TiffData` elements.
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: This sentence is missing a verb and needs reflowing.

"makes for very easy :ref:`TiffData` elements."  easy what?  creation, access, something else?

File: formats/ome-tiff/file-structure.txt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+One of main downside of splitting OME-TIFF over multiple files is their
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: "One of main" should be "The main"

File: formats/ome-tiff/file-structure.txt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+file copying which have the potential to "break" the fileset.
+
+File size
+---------
+
+The TIFF file format internally uses 32-bit byte offsets. The largest offset
+which can be represented is 4GB, making this value the upper limit of the file
+size supported by the design.
+
+The OME-TIFF file format supports the BigTIFF file extensions allowing to
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: "to use" should be "use of"

File: formats/ome-tiff/file-structure.txt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+The main limitation of the BigTIFF file extension is the degree of its
+adoption in the community. Although OME Bio-Formats and OMERO will handle
+OME-TIFF using the BigTIFF file extension, other tools might not be able to
+open it.
+
+Metadata redundancy
+-------------------
+
+Storing multiple planes per OME-TIFF file and keeping the OME-XML metadata
+embedded in every file header is recommended but it is not always practical.
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: "it is" should be "is"

File: formats/ome-tiff/file-structure.txt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+Normally, the OME-XML metadata block is small in comparison with the binary
+pixel data in the file, but in some cases, it may be proportionately larger.
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: disproportionally?

File: formats/ome-tiff/file-structure.txt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+- having many structured annotations in the OME-XML.
+
+For example, if you have a dataset with 1,000 time points, with each plane
+recorded at 512x512 as uint8 pixel type, storing each plane in its own file
+uncompressed requires ~256KB of disk per file, and ~256MB total. But if you
+have 5MB of corresponding OME-XML metadata, embedding a copy of that metadata
+in every file would result in a dataset nearly 20 times larger than before,
+requiring ~5.5MB of disk per file, and more than 5GB total.
+
+One of the advantages of reproducing the OME-XML metadata across file is
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: across files

File: formats/ome-tiff/file-structure.txt [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+redundancy. If even one of the constituent files survives a data crash, the
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: "a data crash" or "data loss"

File: formats/ome-tiff/file-structure.txt (hunk 1 above)
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: information to be easily appended

File: formats/ome-tiff/file-structure.txt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+Since the :doc:`2011-06 version</schemas/june-2011>` of the OME-XML schema, it
+is possible to store partial OME-XML metadata blocks in some or all of the
+TIFF files pointing to a master file containing the full OME-XML metadata.
+The master file can be either another OME-TIFF file or a companion OME-XML
+file.
+
+Using a companion OME-XML file allows to easily append information that caz
+only be generated at the end of the acquisition without the need to manipulate
+existing TIFF IFDs. It has the same drawbacks as the ones highlighted above in
```
Comment (ghost) [inline formats/ome-tiff/file-structure.txt]: s;the ones;those;

File: formats/ome-tiff/specification.txt [hunk 10]
```diff
@@ -1,26 +1,27 @@
...
 `TIFF <http://partners.adobe.com/public/developer/tiff/index.html>`_ and
 :schema_plone:`OME-XML <>` formats, although there is some review of both.
 
-Storing OME-XML within a TIFF
------------------------------
+An OME-TIFF dataset consists of;
+
+- one or more files in standard TIFF format with the file extension
+  ``.ome.tif`` or ``.ome.tiff`` or `BigTIFF format <http://bigtiff.org/>`_
+  with the file extension ``.ome.tf2``, ``.ome.tf8`` or ``.ome.btf``
```
Comment (ghost) [inline formats/ome-tiff/specification.txt]: Note that 'ome.tif' and 'ome.tiff' are also valid when using big tiff; we might want to clarify that here.

File: formats/ome-tiff/specification.txt [hunk 11]
```diff
@@ -1,26 +1,27 @@
...
+- a string of OME-XML metadata embedded in the ImageDescription tag of each
```
Comment (ghost) [inline formats/ome-tiff/specification.txt]: "of each file first IFD" should be "the first IFD of each file"

File: formats/ome-tiff/specification.txt [hunk 12]
```diff
@@ -1,26 +1,27 @@
...
+  file first IFD (Image File Directory).
 
-An OME-TIFF dataset consists of one or more files in standard TIFF or
-`BigTIFF format <http://bigtiff.org/>`_, with the file extension
-``.ome.tif`` or ``.ome.tiff``, and an identical (or in the case of
-multiple files, nearly identical) string of OME-XML metadata embedded
-in the ImageDescription tag of each file's first IFD (Image File
-Directory).  BigTIFF file extensions are also permitted, with the file
-extension ``.ome.tf2``, ``.ome.tf8`` or ``.ome.btf``, but note these
+BigTIFF file extensions are also permitted, with the file extension ``.ome.tf2``, ``.ome.tf8`` or ``.ome.btf``, but note these
```
Comment (ghost) [inline formats/ome-tiff/specification.txt]: This is partially duplicating the detail above; the file extensions could be dropped.

**Comment by ghost** [comment]
The overall intent looks fine.  I've noted a few corrections for readability.

File: formats/ome-tiff/file-structure.txt [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+OME-TIFF filesets distributed over multiple files.
+
+Single versus multiple files
+----------------------------
+
+Splitting a fileset across multiple files can have advantages in terms of
+acquisition but also processing purposes. The OME-TIFF file format can support
+any image organization. However, using one TIFF file per timepoint per channel
+with the focal planes for that timepoint and channel stored sequentially
+within the TIFF makes for very easy :ref:`TiffData` elements.
```
Comment (hflynn) [inline formats/ome-tiff/file-structure.txt]: Just this to clarify now.

Comment (sbesson) [inline formats/ome-tiff/file-structure.txt]: Pushed. As I don't have data to back up any other claim, I think easy creation is the safe call here.

**Comment by sbesson** [comment]
Pushed. One remaining question is the addition of the huge `.ome.xml` dumps. Is there any strong feeling about this ?

**Comment by hflynn** [comment]
Does having the entire xml file make maintenance easier? I don't have a strong objection to it if they are serving a purpose, it's nice to have a clear idea of where inserted content comes from.

**Comment by ghost** [comment]
Thanks, the ome-xml comment looks fine.

Regarding the samples, so long as they are serving a purpose I have no objection.  If they are human-editable, would it make sense to but start and end markers in XML comments so that we can source ranges of lines out of the files for examples without using absolute line numbers.  It would make the documentation more robust in the face of edits to the files.

**Comment by sbesson** [comment]
Regarding the OME-XML content, I simply generated the files using `tiffcomment tubhiswt_C0_TP1.ome.xml > tubhiswt_C0_TP1.ome.xml`. I completely agree with the robustness concerns raised by @rleigh-dundee. Two additional comments:
-  it is unclear to me whether adding Sphinx markers might also reduce the readability of the OME-XML metadata block. We will likely have to try and see if it works.
- it is also interesting to note that these OME-TIFF samples do not illustrate properly the use case discussed in https://github.com/openmicroscopy/ome-documentation/pull/1459#discussion_r61453577. Despite being saved as one Z-stack per file, each plane is described using its own `TiffData` element which is fully correct from a specification point of view but suboptimal. 

At this point, I am inclined to preserve the statu quo, i.e. re-embed the OME-XML sections inline and review this OME-XML metadata block as we tackle the OME-TIFF sample files generation for the upcoming schema (https://trello.com/c/gskA0JKM/65-schema-samples-generation). Thoughts?

**Comment by hflynn** [comment]
Re-embedding the sections inline and reviewing as part of the wider effort works for me.

File: formats/ome-tiff/file-structure.txt [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+For example, if you have a dataset with 1,000 time points, with each plane
+recorded at 512x512 as uint8 pixel type, storing each plane in its own file
+uncompressed requires ~256KB of disk per file, and ~256MB total. But if you
+have 5MB of corresponding OME-XML metadata, embedding a copy of that metadata
+in every file would result in a dataset nearly 20 times larger than before,
+requiring ~5.5MB of disk per file, and more than 5GB total.
+
+One of the advantages of reproducing the OME-XML metadata across files is
+redundancy. If even one of the constituent files survives data loss, the
```
Comment (mtbc) [inline formats/ome-tiff/file-structure.txt]: not quite sure what "even" does here, "one" isn't exactly staggeringly many

**Comment by mtbc** [comment]
I'm left wondering if there's a middle ground of duplicating the full metadata for redundancy but not into every file.

File: formats/ome-tiff/specification.txt [hunk 15]
```diff
@@ -1,25 +1,26 @@
...
 
-
 The following provides technical details on the OME-TIFF
 format. It assumes familiarity with both the
 `TIFF <http://partners.adobe.com/public/developer/tiff/index.html>`_ and
 :schema_plone:`OME-XML <>` formats, although there is some review of both.
 
-Storing OME-XML within a TIFF
------------------------------
+An OME-TIFF dataset consists of;
```
Comment (mtbc) [inline formats/ome-tiff/specification.txt]: intended colon?

File: formats/ome-tiff/specification.txt [hunk 16]
```diff
@@ -285,222 +292,141 @@ reverse temporal order.
...
-stored sequentially within the TIFF) makes for very easy TiffData elements.
+   .. note::
+     The :schema_plone:`FileName <Documentation/Generated/OME-2015-01/ome_xsd.html#TiffData_TiffData_UUID_FileName>`
+     attribute of the UUID is optional, but strongly recommended—otherwise,
+     the OME-TIFF reader must scan OME-TIFF files in the working directory
+     looking for matching UUID signatures.
 
-Multi-file example
-^^^^^^^^^^^^^^^^^^
+When splitting an OME-TIFF across multiple files, the OME-XML metadata must either be embedded into each TIFF file or use partial metadata blocks.
```
Comment (mtbc) [inline formats/ome-tiff/specification.txt]: rather long line

File: formats/ome-tiff/specification.txt [hunk 17]
```diff
@@ -285,222 +292,141 @@ reverse temporal order.
...
-    <OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2015-01"
-        ...
+The only difference between the OME-XML metadata block per TIFF file is the
+:schema_plone:`UUID <Documentation/Generated/OME-2015-01/ome_xsd.html#OME_UUID>`
+attribute of the root OME element. This value should be a distinct
+UUID value for each file, so that each TiffData element can
+unambiguously reference its relevant file using a UUID child element.
 
-::
+The :ref:`tubhiswt_samples` demonstrate how OME-TIFF datasets can be distributed across multiple files. Each of the files in the set has identical metadata apart from the `UUID`, the unique identifier in each file. For example the identifiers could be distributed as follows:
```
Comment (mtbc) [inline formats/ome-tiff/specification.txt]: rather long line

File: formats/ome-tiff/specification.txt [hunk 18]
```diff
@@ -285,222 +292,141 @@ reverse temporal order.
...
-- either store the OME-XML metadata in a companion file with the extension
-  ``.companion.ome`` (see the :ometiff_downloads:`companion fileset sample <companion>`)
-- or choose one OME-TIFF file to serve as the "master" file containing the
-  metadata (see the :ometiff_downloads:`binary only OME-TIFF fileset sample <binaryonly>`)
+Partial OME-XML metadata
+^^^^^^^^^^^^^^^^^^^^^^^^
 
-In the remaining OME-TIFF files where you do not want to store the metadata,
-use an OME-XML block similar to::
+Instead of embedding the full OME-XML metadata into the header each OME-TIFF,
```
Comment (mtbc) [inline formats/ome-tiff/specification.txt]: +of

File: formats/ome-tiff/specification.txt [hunk 19]
```diff
@@ -285,222 +292,141 @@ reverse temporal order.
...
+                   UUID="urn:uuid:4978087c-a670-4b12-af53-256c62d8d101"/>
+    </OME>
 
-    <?xml version="1.0" encoding="UTF-8"?>
-    <OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2015-01"
-        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-        xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2015-01
-            http://www.openmicroscopy.org/Schemas/OME/2015-01/ome.xsd"
-        UUID="urn:uuid:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx">
+The :schema_plone:`MetadataFile <Documentation/Generated/OME-2015-01/ome_xsd.html#OME_OME_BinaryOnly_MetadataFile>` element should contain the name of the master file containing the full
```
Comment (mtbc) [inline formats/ome-tiff/specification.txt]: rather long line

**Comment by sbesson** [comment]
Pushed minor fixes. Re https://github.com/openmicroscopy/ome-documentation/pull/1459#issuecomment-216830821, note also that Curtis proposed a similar structure in the ome-devel thread.

**Comment by mtbc** [comment]
Those fixes look good, thank you.

**Comment by hflynn** [comment]
You all done with this now @sbesson ? Shall I merge and release?

**Comment by sbesson** [comment]
@hflynn: merging would be okay with me assuming everyone is happy with the current state. Let's discuss the release now as @joshmoore might want to review/amend this PR with another one.

**Comment by hflynn** [comment]
Ok, no rush on the release front, I'm hoping to get time to look at https://trello.com/c/3FGRoZZo/422-invalid-ome-tiff before the end of the week, which I hoped we could partly use to point people at this docs update.

**Comment by hflynn** [comment]
Will hold off on merging to give @joshmoore a chance to look

**Comment by hflynn** [comment]
Will merge to take this off Josh's list just now and let us sort out branches, and we can discuss when to release it on the trello card for the related blog post.

**Comment by joshmoore** [comment]
A late :thumbsup:. My only comments can likely be carded as RFEs:
- samples in fileset structure -- having links from the new page would also be useful.
- images -- I've started a graffle with schematics of the different documents; these would be useful on the new page (longer-term a visual description could be added the the sample download pages with a link back to the fileset structure page)
- policy on other tags -- one issue which was in the Micro-Manager thread which I didn't see (but may have missed) is our policy on other IFDs. This could also include the proposal of using annotations as a link (though that should include a full test from our side)

**Comment by sbesson** [comment]
Thanks for the late review. Couple of immediate answers:
- samples in fileset structure - definitely a good thing to add can be opened as a follow-up PR immediately
- :+1: for a explanatory graffle at the top of the structure page
- the IFD policy omission was deliberate to restrict the scope of this PR and give it a chance to be merged ;). There is an upcoming documentation effort on policy/best practices which could include the IFD but also revamping the section under https://www.openmicroscopy.org/site/support/ome-model/developers/using-ome-xml.html. We might want to coordinate with the proposed blog post  discussing similar questions /cc @hflynn @melissalinkert

**Pull request merged**